fastai/callbacks/tensorboard.py

```diff
@@ -0,0 +1,399 @@
+        tag = self.metrics_root + name
+        self.tbwriter.add_scalar(tag=tag, scalar_value=scalar_value, global_step=iteration)
+
+    #TODO:  Relying on a specific hardcoded start_idx here isn't great.  Is there a better solution?
+    def _write_metrics(self, iteration:int, last_metrics:MetricsList, start_idx:int=2)->None:
+        "Writes training metrics to Tensorboard."
+        recorder = self.learn.recorder
+        for i, name in enumerate(recorder.names[start_idx:]):
+            if len(last_metrics) < i+1: return
+            scalar_value = last_metrics[i]
+            self._write_scalar(name=name, scalar_value=scalar_value, iteration=iteration)
+
+    def on_batch_end(self, last_loss:Tensor, iteration:int, **kwargs)->None:
+        "Callback function that writes batch end appropriate data to Tensorboard."
+        if iteration == 0: return
+        self._update_batches_if_needed()
+        if iteration % self.loss_iters == 0: self._write_training_loss(iteration=iteration, last_loss=last_loss)
+        if iteration % self.hist_iters == 0: self._write_weight_histograms(iteration=iteration)
+
+    # Doing stuff here that requires gradient info, because they get zeroed out afterwards in training loop
+    def on_backward_end(self, iteration:int, **kwargs)->None:
+        "Callback function that writes backward end appropriate data to Tensorboard."
+        if iteration == 0: return
+        self._update_batches_if_needed()
+        if iteration % self.stats_iters == 0: self._write_model_stats(iteration=iteration)
+
+    def on_epoch_end(self, last_metrics:MetricsList, iteration:int, **kwargs)->None:
+        "Callback function that writes epoch end appropriate data to Tensorboard."
+        self._write_metrics(iteration=iteration, last_metrics=last_metrics)
+
+# TODO:  We're overriding almost everything here.  Seems like a good idea to question that ("is a" vs "has a")
+class GANTensorboardWriter(LearnerTensorboardWriter):
+    "Callback for GANLearners that writes to Tensorboard.  Extends LearnerTensorboardWriter and adds output image writes."
+    def __init__(self, learn:GANLearner, base_dir:Path, name:str, loss_iters:int=25, hist_iters:int=500, stats_iters:int=100, visual_iters:int=100):
+        super().__init__(learn=learn, base_dir=base_dir, name=name, loss_iters=loss_iters, hist_iters=hist_iters, stats_iters=stats_iters)
+        self.visual_iters = visual_iters
+        self.img_gen_vis = ImageTBWriter()
+        self.gen_stats_updated = True
+        self.crit_stats_updated = True
+
+    def _write_weight_histograms(self, iteration:int)->None:
+        "Writes model weight histograms to Tensorboard."
+        generator, critic = self.learn.gan_trainer.generator, self.learn.gan_trainer.critic
+        self.hist_writer.write(model=generator, iteration=iteration, tbwriter=self.tbwriter, name='generator')
+        self.hist_writer.write(model=critic, iteration=iteration, tbwriter=self.tbwriter, name='critic')
```

Comment:
Vertically align with previous line to make it nicer (iteration, tbwriter and name)